### ISSUE TITLE
Dynamic loss function (changes over generations)

### ISSUE THREAD

**Issue by Jgmedina95** [MilesCranmer/SymbolicRegression.jl#162]
Hi Miles!

Most of this idea builds upon the one discussed here: 
https://github.com/MilesCranmer/SymbolicRegression.jl/issues/92#issuecomment-1103198837
But it is different so I decided to make a new issue. 

Ive been using the idea of 'custom_loss_functions' and it has worked great so far. Its really problem specific so I haven't made any pull requests as I dont see it as a generalizable idea, but this one might be.

Is it possible to make a loss function with a variable hyperparamenter? That changes after X generations? 
like 

`return L2loss + alfa*custom_loss`
so that alfa changes over generations (maybe decreasingly or increasingly) or this would have to be fixed?
 
So far I've been saving the state and restarting it changing the loss parameters on my needs but I was wondering if this could be done since the beginning.

**Comment by MilesCranmer** [comment]
Very interesting idea, I could definitely see this being useful for regularizations!

One difficult thing about implementing this is that the loss function is recalculated only when expressions change. So some of the losses may be out-of-date, especially in the hall of fame.

Do you want the absolute loss to change, or just for the search to favor different things over time? If the latter, it will be much easier. You could modify this line: https://github.com/MilesCranmer/SymbolicRegression.jl/blob/a093714e9416b9ccc0678a48c566948908a9708e/src/Population.jl#L103, which is the loss used in tournaments, with your custom loss. Then the search will favor expressions based on your time variable loss, but the loss actually stored will be fixed to e.g., `L2Loss`.

**Comment by Jgmedina95** [comment]
I got time to think about this:
"One difficult thing about implementing this is that the loss function is recalculated only when expressions change. So some of the losses may be out-of-date, especially in the hall of fame."
But I guess its possible to reevaluate the population members with the new loss function. like 
`for member in populations
member.loss = EvaluateLoss(X, member, new_options)
end`
and similarly with the hall of fame.

The idea could (?) be used to:  1) pause the search, 2) change the loss metric, 3)reevaluate the loss of each member in the saved_state, 4) reevaluate the scores, 5)and then continue. 
Ive got issues with step 4. I thought that using the built-in functions like https://github.com/MilesCranmer/SymbolicRegression.jl/blob/9034341558382188b6c4a9092d847ca8a7162b1d/src/LossFunctions.jl#L50
would help me reevaluate all the scores, but when I try to replicate the scores of a member I don't get the same value.

![image](https://user-images.githubusercontent.com/97254349/206051062-fc70ece4-bedf-40c0-87bc-ab77da07cade.png)

If all this was possible, a dynamic loss could be done externally without changing too much into the code, and just readjusting the population, as described, accordingly right?

And to answer the question, yes, im more interested in adjusting the search to favor different things :)

**Comment by MilesCranmer** [comment]
So the proper way to solve this is to run for n iterations, and then change the loss function (e.g., by editing `mach.model.elementwise_loss`), and do a warm-start. This can be repeated.
